### PR TITLE
AP_HAL_ChibiOS: Optimizations for external flash targets (H750 etc.)

### PR DESCRIPTION
**Pull request by mluessi**
I have been working on support for a new target (BrainFPV RADIX 2 HD), which is STM32H750 based. The target uses a custom bootloader and the code is stored in an external flash connected via QUADSPI running at 120 MHz.

I found that the CPU load was quite high, about 80% with the default settings for Copter. Some of the time critical code had already been relocated to RAM. To optimize it further, I used a logic analyzer to capture the QUADSPI traffic and then analyzed it to see what functions / symbols are read most frequently within a 1 second period. The number of QUADSPI transactions in 1 second was about 1.1 million, which is quite impressive. I captured data repeatedly and moved functions and symbols to RAM (ITCM, DTCM, flashram). By doing so, I was able to reduce the CPU load for Copter to about 40%.  

Among the symbols I relocated is the vector table. This needed some changes to the makefile, hwdef script, and linker files so the ChibiOS startup script copies the vector table and sets the VTOR register to the new address.

There are some other changes that were needed to make the target work. Namely, the option to set the QUADSPI clock source and force the STM32 clock initialization. This is needed because the target uses a custom bootloader and the clock configuration the bootloader uses is different than in ArduPilot. By using the HCLK for the QUADSPI, the clock can be reconfigured even though the code is running from external flash. 

To summarize the changes:
- Copy the vector table to RAM if the code is running from external flash
- Move time critical functions to RAM for H750. The same optimizations should also be applied to other external flash targets, but this hasn't been done yet because I don't have hardware to test
- Allow setting the QUADSPI clock source and forcing the STM32 clock initialization
- Increase the flashram for H750 from 256k to 384k
- Fix CAN pin naming in the H750 script (was FDCAN instead of CAN)

I have bench tested this for Copter and Plane and also "hover tested" it for Copter.

### CODE REVIEW

File: libraries/AP_HAL_ChibiOS/hwdef/common/common_extf_h730.ld [hunk 1]
```diff
@@ -74,10 +74,14 @@ SECTIONS
...
-    startup : ALIGN(16) SUBALIGN(16)
+    /* Vector table: copied to RAM by startup script */
+    vectors : ALIGN(1024) SUBALIGN(16)
     {
+        __textvectors_base__ = LOADADDR(vectors);
+        __vectors_base__ = .;
         KEEP(*(.vectors))
-    } > default_flash
+        __vectors_end__ = .;
+    } > flashram AT > default_flash
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/common/common_extf_h730.ld:84]: You need to try building this with the H7RF - it has less RAM and this may not fit.

Wouldn't the vector table be better in DTCM?

File: libraries/AP_HAL_ChibiOS/hwdef/common/common_extf_h750.ld [hunk 2]
```diff
@@ -249,13 +266,13 @@ SECTIONS
     .ARM.extab :
     {
         *(.ARM.extab* .gnu.linkonce.armextab.*)
-    } > default_flash
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/common/common_extf_h750.ld:252]: a comment about this would be good

File: libraries/AP_HAL_ChibiOS/hwdef/common/common_extf_h750.ld [hunk 3]
```diff
@@ -249,13 +266,13 @@ SECTIONS
...
+    } > instram AT > default_flash
 
     .ARM.exidx : {
         __exidx_start = .;
         *(.ARM.exidx* .gnu.linkonce.armexidx.*)
         __exidx_end = .;
-     } > default_flash
+     } > instram AT > default_flash
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/common/common_extf_h750.ld:275]: comment here as well

File: libraries/AP_HAL_ChibiOS/hwdef/common/board.c [hunk 4]
```diff
@@ -236,7 +236,7 @@ void __early_init(void) {
 #if !defined(STM32F1)
   stm32_gpio_init();
 #endif
-#if !HAL_XIP_ENABLED
+#if !HAL_XIP_ENABLED || defined(STM32_FORCE_CLOCK_INIT)
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/common/board.c:239]: This is our option so I think needs to be called HAL_FORCE_CLOCK_INIT

**Review by andyp1per — CHANGES_REQUESTED** [review]
Very nice, some minor comments

File: libraries/AP_HAL_ChibiOS/hwdef/common/common_extf_h730.ld (hunk 1 above)
Comment (mluessi) [reply libraries/AP_HAL_ChibiOS/hwdef/common/common_extf_h730.ld:84]: @andyp1per OK, yes, DTCM should be even better. Let me verify that it works

File: libraries/AP_HAL_ChibiOS/hwdef/common/common_extf_h750.ld (hunk 2 above)
Comment (mluessi) [reply libraries/AP_HAL_ChibiOS/hwdef/common/common_extf_h750.ld:252]: Will do. It needs to be in RAM because libgcc.a was moved there

**Comment by mluessi** [comment]
@andyp1per I made the changes you suggested. 

The RAM of the SPRacingH7RF target is quite full, but it still fits (for Copter):
```
  Memory region         Used Size  Region Size  %age Used
   default_flash:          2 MB         2 MB    100.00%
         instram:       62312 B        63 KB     96.59%
            ram0:        272 KB       272 KB    100.00%
        flashram:       32028 B        32 KB     97.74%
         dataram:       54208 B        64 KB     82.71%
```

**Review by andyp1per — APPROVED** [review]
Very nice!

File: libraries/AP_HAL_ChibiOS/hwdef/common/common_extf_h750.ld [hunk 5]
```diff
@@ -212,6 +228,7 @@ SECTIONS
         lib/lib*.a:vector3.*(.rodata*)
         lib/lib*.a:matrix3.*(.rodata*)
         *libm.a:*(.rodata*)
+        ArduCopter/Copter.*(.rodata*)
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/common/common_extf_h750.ld:231]: I think this is for the scheduler table, what about AP_Vehicle?

Comment (mluessi) [reply libraries/AP_HAL_ChibiOS/hwdef/common/common_extf_h750.ld:231]: Yes, it's for the scheduler table. It would probably make sense to add all of them and group them together. Does the code below look correct to you?

```
AntennaTracker/Tracker.*(.rodata*)
AP_Vehicle/AP_Vehicle.*(.rodata*)
ArduCopter/Copter.*(.rodata*)
ArduPlane/ArduPlane.*(.rodata*)
ArduSub/ArduSub.*(.rodata*)
Blimp/Blimp.*(.rodata*)
Rover/Rover.*(.rodata*)
```

Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/hwdef/common/common_extf_h750.ld:231]: Makes sense to me - with a comment

Comment (mluessi) [reply libraries/AP_HAL_ChibiOS/hwdef/common/common_extf_h750.ld:231]: @andyp1per WDYT? Should I add the above change?

Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/hwdef/common/common_extf_h750.ld:231]: yes

Comment (mluessi) [reply libraries/AP_HAL_ChibiOS/hwdef/common/common_extf_h750.ld:234]: > yes

Done :)

**Comment by mluessi** [comment]
@andyp1per thanks for the review. Please let me know if there is anything I can do to get this merged.

**Comment by andyp1per** [comment]
Tested on my SPRacingH7

**Comment by tridge** [comment]
@mluessi very nice work, thanks!

**Comment by peterbarker** [comment]
Merged, thanks!

**Comment by mluessi** [comment]
Thanks! Great to see that it was merged :)

**Comment by mluessi** [comment]
@tridge would it be possible to backport this to 4.4? It would be nice to have support for the target I have been working on (see #23948) in the 4.4 release

**Comment by andyp1per** [comment]
You will also need https://github.com/ArduPilot/ChibiOS/commit/26e3904be651900477de901c52ecb5dd33021057 backported

**Comment by mluessi** [comment]
> You will also need [ArduPilot/ChibiOS@26e3904](https://github.com/ArduPilot/ChibiOS/commit/26e3904be651900477de901c52ecb5dd33021057) backported

 Looks like a critical bug to me, so backporting that seems like a good idea.

**Comment by peterbarker** [comment]
@mluessi sadly this PR seems to have killed SPRacingH7 bootloader build in an interesting fashion.  I bsected it to a single commit:
```
pbarker@fx:~/rc/ardupilot((no branch, bisect started on da95985dd7c))$ git bisect good
0c93800ec1d3b1a3b4d1fb7358f0acf864d4b385 is the first bad commit
commit 0c93800ec1d3b1a3b4d1fb7358f0acf864d4b385
Author: Martin Luessi <mluessi@gmail.com>
Date:   Thu Jun 1 08:50:09 2023 -0700

    AP_HAL_ChibiOS: Copy vector table to RAM if external flash is primary

 libraries/AP_HAL_ChibiOS/hwdef/common/chibios_board.mk  | 4 ++++
 libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py | 1 +
 2 files changed, 5 insertions(+)
```

Command I used for the bisect:
```
rm -rf build && git submodule update --init --recursive && ./waf configure --board=SPRacingH7 && ./waf copter && ./waf configure --board=SPRacingH7 --bootloader && ./waf bootloader
```

Error message from that when it fails:
```
[101/106] Linking build/SPRacingH7/lib/libAP_Bootloader_libs.a
[102/106] Linking build/SPRacingH7/bootloader/AP_Bootloader
Memory region         Used Size  Region Size  %age Used
           flash:        128 KB       128 KB    100.00%
            ram0:        128 KB       128 KB    100.00%

/home/pbarker/gcc/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: modules/ChibiOS//libch.a(crt0_v7m.o): in function `endfiniloop':
(.text+0x114): undefined reference to `__textvectors_base__'
/home/pbarker/gcc/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: (.text+0x118): undefined reference to `__vectors_base__'
/home/pbarker/gcc/gcc-arm-none-eabi-10-2020-q4-major/bin/../lib/gcc/arm-none-eabi/10.2.1/../../../../arm-none-eabi/bin/ld: (.text+0x11c): undefined reference to `__vectors_end__'
collect2: error: ld returned 1 exit status
```

( @mluessi )

**Comment by andyp1per** [comment]
I don't think there is anything wrong with this PR - its the age old problem of chibios not detecting changes. In particular it doesn't detect changes to the linker script. Builds fine after clean
